### PR TITLE
Remodel SyncManager as a continuously running service

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Changes `SyncManager` into a service which is expected to continue keeping us in sync as long as it's running, rather than a one-off event to bring us up to sync.  It's actual behaviour is mostly unchanged - it does an initial sync and triggers a new sync whenever a new peer connects.  The two key differences are:
* Exceptions thrown during a sync are logged instead of silently lost because we didn't handle the returned `CompletableFuture`
* If a peer connects while an existing sync is in progress, we schedule a new sync for when that one completes. This ensures that if the peer has a better chain we will eventually get there.

Net result is that we keep running syncs until there is no longer a peer with a better finalised epoch.

`PeerSync` got a bit of reworking so that `SyncManager` could be tested independently of it.  It now accepts the peer to sync to as an argument to the `sync` method rather than the constructor so it can be re-used for multiple syncs.

### CODE REVIEW

File: sync/src/main/java/tech/pegasys/artemis/sync/SyncService.java [hunk 1]
```diff
@@ -30,20 +30,19 @@ public SyncService(
       final Eth2Network network,
       final ChainStorageClient storageClient,
       final BlockImporter blockImporter) {
-    this.syncManager = new SyncManager(network, storageClient, blockImporter);
+    this.syncManager =
+        new SyncManager(network, storageClient, new PeerSync(storageClient, blockImporter));
```
Comment (mbaxter) [inline sync/src/main/java/tech/pegasys/artemis/sync/SyncService.java]: (optional) `PeerSync` feels like an internal detail of the `SyncManager`.  What about having a factory method: `SyncManager.create(network, storageClient, blockImporter)` that creates `PeerSync` and passes it to a package-private `SyncManager` constructor?

File: sync/src/test/java/tech/pegasys/artemis/sync/SyncManagerTest.java [hunk 2]
```diff
@@ -77,59 +66,114 @@ public void setUp() {
...
+    assertThat(syncManager.isSyncActive()).isFalse();
+    assertThat(syncManager.isSyncQueued()).isFalse();
+
     verify(network).subscribeConnect(onConnectionListener.capture());
     final PeerConnectedSubscriber<Eth2Peer> subscriber = onConnectionListener.getValue();
 
-    final CompletableFuture<Void> requestFuture = CompletableFuture.completedFuture(null);
+    final CompletableFuture<PeerSyncResult> syncFuture1 = new CompletableFuture<>();
+    final CompletableFuture<PeerSyncResult> syncFuture2 = new CompletableFuture<>();
+    when(peerSync.sync(peer)).thenReturn(syncFuture1).thenReturn(syncFuture2);
```
Comment (mbaxter) [inline sync/src/test/java/tech/pegasys/artemis/sync/SyncManagerTest.java]:
```suggestion
    when(peerSync.sync(peer)).thenReturn(syncFuture1);
```

File: sync/src/main/java/tech/pegasys/artemis/sync/SyncManager.java [hunk 3]
```diff
@@ -13,60 +13,102 @@
...
 
-  private CompletableFuture<Void> executeSync() {
-    Optional<Eth2Peer> possibleSyncPeer = findBestSyncPeer();
+  @Override
+  protected CompletableFuture<?> doStop() {
+    network.unsubscribeConnect(peerConnectSubscriptionId);
+    synchronized (this) {
+      syncQueued = false;
+    }
+    return completedFuture(null);
```
Comment (mbaxter) [inline sync/src/main/java/tech/pegasys/artemis/sync/SyncManager.java:69]: Should we make sure our last request completes or is cancelled?  I think it's possible that we could keep sending out requests if there's an outstanding request that completes unsuccessfully and causes us to retry, potentially multiple times.

**Review by mbaxter — APPROVED** [review]
Looks good!  Just wondering about the shutdown logic

File: sync/src/main/java/tech/pegasys/artemis/sync/SyncService.java (hunk 1 above)
Comment (ajsutton) [reply sync/src/main/java/tech/pegasys/artemis/sync/SyncService.java]: Yep, makes sense. Done.

File: sync/src/main/java/tech/pegasys/artemis/sync/SyncManager.java (hunk 3 above)
Comment (ajsutton) [reply sync/src/main/java/tech/pegasys/artemis/sync/SyncManager.java:69]: Good spot. Done.

**Pull request merged**